src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/MapCode/AbstractMapCodeService.cs

```diff
@@ -0,0 +1,356 @@
+    {
+        // We create a new Razor file based on each content in each mapping order to get the syntax tree that we'll later use to map.
+        var newSnapshot = documentContext.Snapshot.WithText(SourceText.From(content));
+        var codeToMap = await newSnapshot.GetGeneratedOutputAsync(cancellationToken).ConfigureAwait(false);
+        var syntaxTree = codeToMap.GetSyntaxTree();
+        if (syntaxTree is null)
+        {
+            return [];
+        }
+
+        var nodesToMap = ExtractValidNodesToMap(syntaxTree.Root);
+
+        return nodesToMap;
+    }
+
+    private async Task<CSharpFocusLocationsAndNodes?> GetCSharpFocusLocationsAndNodesAsync(ISolutionQueryOperations queryOperations, ImmutableArray<RazorSyntaxNode> nodesToMap, LspLocation[][] locations, CancellationToken cancellationToken)
+    {
+        using var csharpNodes = new PooledArrayBuilder<string>();
+        foreach (var nodeToMap in nodesToMap)
+        {
+            if (nodeToMap.IsCSharpNode(out var csharpBody))
+            {
+                csharpNodes.Add(csharpBody.ToString());
+            }
+        }
+
+        if (csharpNodes.Count > 0)
+        {
+            var csharpFocusLocations = await GetCSharpFocusLocationsAsync(queryOperations, locations, cancellationToken).ConfigureAwait(false);
+
+            return new CSharpFocusLocationsAndNodes(csharpFocusLocations, csharpNodes.ToArray());
+        }
+
+        return null;
+    }
+
+    private static async Task TryMapRazorNodesAsync(DocumentContext documentContext, string content, LspLocation[][] focusLocations, List<TextDocumentEdit> changes, CancellationToken cancellationToken)
+    {
+        var nodesToMap = await GetNodesToMapAsync(documentContext, content, cancellationToken).ConfigureAwait(false);
+        if (nodesToMap.IsDefaultOrEmpty)
+        {
+            // No nodes to map or syntax root is null, nothing to do.
+            return;
+        }
+
+        var syntaxTree = await documentContext.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+        var syntaxRoot = syntaxTree.Root;
+
+        foreach (var locationByPriority in focusLocations)
+        {
+            foreach (var location in locationByPriority)
+            {
+                Debug.Assert(location.DocumentUri.GetRequiredParsedUri() == documentContext.Uri);
+
+                using var razorNodesToMap = new PooledArrayBuilder<RazorSyntaxNode>();
+                foreach (var nodeToMap in nodesToMap)
+                {
+                    // Only process new, non-C#, nodes
+                    if (nodeToMap.IsCSharpNode(out _) || nodeToMap.ExistsOnTarget(syntaxRoot))
+                    {
+                        continue;
+                    }
+
+                    razorNodesToMap.Add(nodeToMap);
+                }
+
+                var sourceText = await documentContext.Snapshot.GetTextAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
Isn't this the same as `syntaxTree.Source.Text`? Seems wasteful to await an async method to get the same result in these loops.